gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/MessageServiceImpl.java

```diff
@@ -335,6 +311,34 @@ else if (api != null && RoleScope.APPLICATION.name().equals(recipientEntity.getR
         }
     }
 
+    private Set<String> getIndirectMemberIds(ExecutionContext context, List<String> applicationIds, RoleEntity roleEntity) {
+        if (!applicationIds.isEmpty()) {
+            // get all indirect members
+            Optional<Set<String>> applicationsGroups = applicationService
+                .findByIds(context, applicationIds)
+                .stream()
+                .map(ApplicationListItem::getGroups)
+                .reduce(
+                    (a, b) -> {
+                        a.addAll(b);
+                        return a;
+                    }
+                );
+            if (applicationsGroups.isPresent()) {
+                return membershipService
+                    .getMembershipsByReferencesAndRole(
+                        MembershipReferenceType.GROUP,
+                        new ArrayList<>(applicationsGroups.get()),
+                        roleEntity.getId()
+                    )
+                    .stream()
+                    .map(MembershipEntity::getMemberId)
+                    .collect(Collectors.toSet());
+            }
+        }
+        return null;
+    }
```

Comment:
I think it can be simplified to ⬇️ (with maybe an extra empty check on `applicationsGroups`):
```java
    private Set<String> getIndirectMemberIds(ExecutionContext context, List<String> applicationIds, RoleEntity roleEntity) {
        if (applicationIds.isEmpty()) {
            return null;
            // OR better
            // return Collections.emptySet();
        }

        // Get all indirect members
        List<String> applicationsGroups = applicationService
            .findByIds(context, applicationIds)
            .stream()
            .flatMap((ApplicationListItem applicationListItem) -> applicationListItem.getGroups().stream())
            .distinct()
            .collect(Collectors.toList());

        return membershipService
            .getMembershipsByReferencesAndRole(MembershipReferenceType.GROUP, applicationsGroups, roleEntity.getId())
            .stream()
            .map(MembershipEntity::getMemberId)
            .collect(Collectors.toSet());
    }
```

Reply:
👍